### ISSUE TITLE
Excessive RAM usage?

### ISSUE THREAD

**Issue by uppfinnarn** [TryGhost/Ghost#6211]
I was looking through the process monitor on one of our servers, and noticed Ghost was taking a whooping 250MB RAM. This seems… rather excessive for something like Ghost, and a whole lot more than what I remember of previous versions.

I checked my personal blog (which barely gets any traffic), and it's taking ~103MB (running on a rather outdated Ghost install, because I kinda forgot it existed - 0.5.0). The other differences are that the 250MB RAM one is using a Postgres database rather than a sqlite one (it's load balanced between two web nodes), and has slightly different email configuration.

What could have caused this increase in RAM usage? Is it caching something in RAM, and can I tell it not to? Is it a bug or a feature?

**Comment by rwboyer** [comment]
I've found that this is pretty much the norm for a while now... at least since 6.0 from my perspective. On average they seem to level out around 320MB per instance for moderate traffic sites (>500 <1000 sessions/day) I've not really looked at it hard but there's most likely some sort of node/library/whatever memory leaks as is the case for most node apps somewhere. It doesn't seem to be devastating and you may be able to impact that with one of the 9 million node start-up parameters having to do with GC behavior. 

On hosting with platforms like heroku, etc it's not a super big deal but can be a giant PIA for hosting small-ish sites on platforms where there is plenty of CPU/bandwidth to host 50 ghost sites but memory bumps cost a ton of money (IE webfaction and other shared accounts/servers).

**Comment by ErisDS** [comment]
About a month ago, some quite significant testing was done on Ghost to try to see if there was a memory leak. As far as that testing was able to distinguish, there is no discernible leak (cc @acburdine do you have the scripts that were used for that?).

That doesn't mean that the memory usage isn't quite high, just that it is stable.

High memory usage can be accounted by a few things. Ghost stores quite a few things in memory, including a cache of the settings table for quick access, and a cache of your templates, so if you have a particularly large or complex theme, that can push it up a bit. Also, the sitemaps are stored in memory, so if you have a large site, that'll also impact the size of the memory footprint. Further more, the way Ghost loads its own internal dependencies means it likely keeps a whole bunch of code in memory that it doesn't need to. 

There are definitely ways and means to optimise and improve Ghost's usage of memory, however my first priority is always that the footprint be stable, rather than small.

It's also worth noting, that Ghost is designed to have one instance with a cache in front, rather than multiple instances - the use of memory storage is the major reason why having multiple instances can lead to unexpected behaviour.

More than happy to hear suggestions for, and see PRs to improve Ghost's memory footprint.

**Comment by halfdan** [comment]
You may want to take a look at this article: https://blog.risingstack.com/finding-a-memory-leak-in-node-js/

TL;DR; The GC doesn't run often on newer versions - node.js tries to use 1.5GB of memory and the GC is rarely triggered before this threshold is reached. Lower the memory to force the GC to run more often.

**Comment by cwonrails** [comment]
What tests are you all using for RAM usage? Ghost is using ~121MB on my VPN at the moment.

**Comment by uppfinnarn** [comment]
Interestingly, just upgrading to Node 4.2 and Ghost 0.7.3 seems to have dropped the exact same setup down to ~80MB, and I can gain about 20-30MB on top of that with `--optimize_for_size` if I want to.

I don't know what Node 4.x did to make this happen, but I like it.

**Comment by acburdine** [comment]
@ErisDS in terms of the scripts that were used for testing, I don't have those unfortunately :disappointed: . I built the script that created test data in a Ghost blog, but @vdemedes was the one who had all of the Docker setup/memory testing scripts.

**Comment by ErisDS** [comment]
No worries, I have them, I was wondering if you had them all setup :wink:

**Comment by ErisDS** [comment]
@uppfinnarn I'm going to close this based on your last comment. There are things which can affect the RAM usage inside of Ghost as I mentioned, but it seems you found a way to get your blog running in a sensible amount of RAM.
